### PR TITLE
Parallelize Silicon::accumulate using OpenMP

### PR DESCRIPTION
**Pull request by jamesp-epcc**
This is my first attempt at adding OpenMP to the photon loop and pixel distortions loop in Silicon::accumulate, as discussed in #1008 . I suspect this code is not yet ready to be merged but it would be good to get feedback on what changes it would need.

### CODE REVIEW

File: src/Silicon.cpp [hunk 1]
```diff
@@ -536,136 +573,163 @@ namespace galsim {
...
-                double diffStep = std::max(
-                    0.0, diffStep_pixel_z * std::sqrt(zconv * _sensorThickness));
-                x0 += diffStep * gd();
-                y0 += diffStep * gd();
-            }
-            xdbg<<" => "<<x0<<','<<y0<<std::endl;
-            double flux = photons.getFlux(i);
+	std::vector<GaussianDeviate> gd;
+	for (int i = 0; i < numThreads; i++) {
+	   gd.push_back(GaussianDeviate(ud[i], 0, 1));
```
Comment (rmjarvis) [inline src/Silicon.cpp]: Lots of places here where you use 3 spaces rather than our standard 4.  These should be fixed.

**Comment by jamesp-epcc** [comment]
Thanks for the feedback. I have made some changes to address this. The random numbers are now generated in the main thread before the photon loop so that the results are consistent across different numbers of threads. However, the results are different from the original code because it's not possible to get the random sequence to be exactly the same (one call to the RNG only happens if the pixel is not found, and this can't be predicted in advance). Therefore I have had to make some changes to `test_silicon_area` to address this.
In addition, one test that previously matched to 15 decimal places now only matches to 14 when running with multiple threads. I think this is probably just due to subtle differences caused by the different order of operations when running in parallel.

**Comment by rmjarvis** [comment]
Differences at the 15th or so decimal place are fine.  You get the same kind of differences when blended galaxies are rendered in a different order and the numerics care which order numbers are added up.  But we do want to avoid differences of order unity when a photon is placed on a different pixel depending on the number of threads.

**Comment by jamesp-epcc** [comment]
I haven't noticed any larger differences. With the updated code and tests, results are extremely close even with different numbers of threads, for all the cases I tried.

File: .travis.yml [hunk 2]
```diff
@@ -62,6 +62,9 @@ script:
 
     - cd tests
 
+    # Running with multiple threads produces different results due to RNG sequence
+    - export OMP_NUM_THREADS=1
```
[the remark this answers is not in the document]
Comment (jamesp-epcc) [reply .travis.yml]: Removed, no longer needed now that new version produces same results with any thread count.

File: setup.py [hunk 3]
```diff
@@ -69,7 +69,7 @@ def all_files_from(dir, ext=''):
     undef_macros+=['NDEBUG']
 
 copt =  {
-    'gcc' : ['-O2','-msse2','-std=c++11','-fvisibility=hidden'],
+    'gcc' : ['-O2','-msse2','-std=c++11','-fvisibility=hidden','-fopenmp'],
```
[the remark this answers is not in the document]
Comment (jamesp-epcc) [reply setup.py]: Added the option for icc and clang.

File: src/Silicon.cpp [hunk 4]
```diff
@@ -157,7 +169,7 @@ namespace galsim {
     void Silicon::updatePixelDistortions(ImageView<T> target)
     {
         dbg<<"updatePixelDistortions\n";
-        // This updates the pixel distortions in the _imagepolys
+       // This updates the pixel distortions in the _imagepolys
```
[the remark this answers is not in the document]
Comment (jamesp-epcc) [reply src/Silicon.cpp]: Fixed

File: src/Silicon.cpp [hunk 5]
```diff
@@ -462,7 +482,24 @@ namespace galsim {
...
+	// Generate RNGs for each thread.
+	std::vector<UniformDeviate> ud;
+	if (numThreads == 1) {
+	    // Match the old code in the serial case
+	    ud.push_back(UniformDeviate(rng));
+	}
+	else {
+	    UniformDeviate initialUd(rng);
+	    for (int i = 0; i < numThreads; i++) {
+		ud.push_back(UniformDeviate(initialUd() * LONG_MAX));
```
[the remark this answers is not in the document]
Comment (jamesp-epcc) [reply src/Silicon.cpp]: I have changed the code to generate the numbers on the main thread and pass them to the other threads for reproducibility.

**Comment by jamesp-epcc** [comment]
Just looked at this again and realised I never marked the change requests as resolved. Apologies, have done that now.

**Pull request merged**